### PR TITLE
Optimize MatchAny

### PR DESCRIPTION
**Pull request by JojiiOfficial**
### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

/claim #3522 

**Note:** in /Cargo.toml:181 I used a custom fork of schemars to allow using SmolStr directly in the API. This is unsecure and should be replaced with a fork in the qdrant organization like the tonic patch above!

# Benches
I created a new bench function `conditional-search-large-match-any` that measures the performance increase. To test the performance of the old implementation one can simply checkout the first commit of this PR.

Results for 100k points, 1k MatchAny values of which 15% can be a matching value:
Old: `1.07ms`
This PR: `561.12µs`

### CODE REVIEW

**Comment by JojiiOfficial** [comment]
I experienced a further increase in performance around 5-15% in most filtering benchmarks using the fnv hashing algorithm instead of the std's SipHash. 
The [fnv crate](https://github.com/servo/rust-fnv) is already part of the dependency tree, so it doesn't add any further crates that need to be compiled when compiling the qdrant binary.

**Comment by generall** [comment]
Hey @JojiiOfficial, thanks for the PR.

Couple of things to consider from my side:

- Would it be possible to avoid usage of non-merged PRs? I understood that's for the SmolStr, but is it really necessary for the internal optimization? Maybe we could decompose external structures from the internal logic?
- We need to make sure, that those changes doesn't compromise the performance of the default use-case with small number of values

**Comment by JojiiOfficial** [comment]
Thanks for the reply @generall 

Indeed I found a way to remove the SmolStr from the external structure. I also added a benchmark for a small number of keyword values and optimized the comparison logic. 

With the latest changes I don't get a real performance regression over multiple benchmarks runs. In contrast most other conditional-search benches got a slight increase in performance in average.

**Comment by generall** [comment]
FYI I was using our benchmarking tool with the following changes: https://github.com/qdrant/bfb/commit/cb1a77e66b4dc7437505165509bbcf6b703bd446

**Comment by JojiiOfficial** [comment]
You're right! Thanks for checking.
I indeed missed a spot where the filter gets checked. As you can see from the benchmark results below, It is nearly twice as fast for 200 keywords per query. Even for only 20 points we have a performance improvement of around 22%
I'll experience and try to optimize a bit more and push my changes later.

image

File: docs/redoc/master/openapi.json [hunk 1]
```diff
@@ -6585,7 +6585,8 @@
             "type": "array",
             "items": {
               "type": "string"
-            }
+            },
+            "uniqueItems": true
```
Comment (coszio) [inline docs/redoc/master/openapi.json:6589]: This might lead to a breaking change in clients, is it possible to remove?

File: lib/segment/src/index/query_optimization/condition_converter.rs [hunk 2]
```diff
@@ -332,9 +336,13 @@ pub fn get_match_checkers(index: &FieldIndex, cond_match: Match) -> Option<Condi
             (AnyVariants::Keywords(list), FieldIndex::KeywordIndex(index)) => {
                 Some(Box::new(move |point_id: PointOffsetType| {
                     index.get_values(point_id).map_or(false, |values| {
-                        values
-                            .iter()
-                            .any(|k| !list.iter().any(|s| s.as_str() == k.as_ref()))
+                        if list.len() < 13 {
```
Comment (coszio) [inline lib/segment/src/index/query_optimization/condition_converter.rs:339]: We should put this magic number in a constant along with a comment to explain where it comes from

File: lib/segment/src/payload_storage/condition_checker.rs [hunk 3]
```diff
@@ -91,15 +91,27 @@ impl ValueChecker for Match {
...
                 (Value::Number(stored), AnyVariants::Integers(list)) => stored
                     .as_i64()
                     .map(|num| list.contains(&num))
                     .unwrap_or(false),
                 _ => false,
             },
             Match::Except(MatchExcept { except }) => match (payload, except) {
-                (Value::String(stored), AnyVariants::Keywords(list)) => !list.contains(stored),
+                (Value::String(stored), AnyVariants::Keywords(list)) => {
+                    if list.len() < 13 {
```
Comment (coszio) [inline lib/segment/src/payload_storage/condition_checker.rs:109]: Same here

File: docs/redoc/master/openapi.json (hunk 1 above)
Comment (generall) [reply docs/redoc/master/openapi.json:6589]: I don't think client enforces this one. I propose to ignore for now

File: lib/segment/src/types.rs [hunk 4]
```diff
@@ -1185,7 +1187,7 @@ pub enum ValueVariants {
 #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq, Eq)]
 #[serde(untagged)]
 pub enum AnyVariants {
-    Keywords(Vec<String>),
+    Keywords(IndexSet<String, FnvBuildHasher>),
     Integers(Vec<IntPayloadType>),
```
Comment (coszio) [inline lib/segment/src/types.rs:1191]: Why are integers not optimized?

**Review by coszio — CHANGES_REQUESTED** [review]

File: lib/segment/src/index/query_optimization/condition_converter.rs (hunk 2 above)
Comment (generall) [reply lib/segment/src/index/query_optimization/condition_converter.rs:339]: +1

File: lib/segment/src/types.rs (hunk 4 above)
Comment (JojiiOfficial) [reply lib/segment/src/types.rs:1191]: My initial benchmark showed that searching integer iteratively were faster than hashing for a reasonable sized list. I assumed this were the case as integer perfectly fit into cache while being very small. 
Doing a new quick bench however showed

Comment (coszio) [reply lib/segment/src/types.rs]: Is the threshold the same for both strings and integers? 

Another quick idea (for both strings and integers) is to do binary search over a sorted list as a replacement of iterating. I bet this can increase the threshold before which hashing is better. Although, to be fair, this would add extra complexity and can be addressed later in another PR.

Comment (timvisee) [reply lib/segment/src/types.rs]: Maybe we can base this threshold on number of bytes instead.

> Another quick idea (for both strings and integers) is to do binary search over a sorted list as a replacement of iterating. I bet this can increase the threshold before which hashing is better. Although, to be fair, this would add extra complexity and can be addressed later in another PR.

Yes, I've been thinking about this too. But it would make our logic a bit more complex as we likely end up with two different search methods based on some heuristic.

**Review by generall — APPROVED** [review]

File: lib/segment/src/types.rs (hunk 4 above)
Comment (coszio) [reply lib/segment/src/types.rs]: I played around with the comparisons between binary search, linear scan, and IndexSet for searching UUIDs. The results favor the current implementation (the one in this PR 🎉).

Before a small threshold, linear scan and binary search perform basically the same. Y-Axis is nanoseconds, X-Axis is list size.
imagen

However maybe the threshold (at least for UUIDs) might be better adjusted at size of 30 items.

TLDR: there is a really small optimization that can be made by setting the threshold based on a byte size heuristic, but I don't think it is worth the complexity.

Comment (timvisee) [reply lib/segment/src/types.rs]: I guess this to be very dependent on hardware configuration.

But what about setting a threshold at 512 bytes (32 * 16 (UUID) = 512)? It might scale nicely that way.
